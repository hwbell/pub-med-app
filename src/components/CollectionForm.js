import React from 'react';
import '../App.scss';
import 'bootstrap/dist/css/bootstrap.css';

// components
import { Button, ButtonGroup, Input, Form, InputGroup, Modal, ModalHeader, ModalBody, ModalFooter } from 'reactstrap';

// functions
import { saveCollection } from '../tools/serverFunctions';

// this form is toggled visible by the 'add to collection' button on the search page
class CollectionForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: '',
      highlightInd: null,
      updating: false,
      creating: false
    }

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.patchCollection = this.patchCollection.bind(this);
    this.combineAllCollections = this.combineAllCollections.bind(this);
    this.renderMessage = this.renderMessage.bind(this);
    this.renderSelectors = this.renderSelectors.bind(this);
    this.renderInput = this.renderInput.bind(this);
    this.renderArticlePreview = this.renderArticlePreview.bind(this);
  }

  componentDidMount() {
    // if there are no collections in props, 
    // check for localStorage Collections and register them in app
    let noPropsCollections = this.props.collections.length === 0;
    let localCollections = JSON.parse(localStorage.getItem('collections'));

    if (localCollections && localCollections.length > 0 && noPropsCollections) {
      localCollections.forEach((collection) => {
        this.props.createNewCollection(collection, () => {
          console.log(`${collection.name} registered from localStorage`)
        });
      })
    }

  }

  // use this function to combine the user collections(if present) with the props.collections(if present)
  combineAllCollections() {

    let propCollections = JSON.parse(JSON.stringify(this.props.collections));
    // console.log(propCollections)

    let allCollections;
    if (this.props.user) {
      let userCollections = JSON.parse(JSON.stringify(this.props.user.collections)) || [];
      // console.log(userCollections)

      allCollections = propCollections.concat(userCollections);
    } else {
      allCollections = propCollections;
    }

    return allCollections;
  }

  highlight(i) {
    if (this.state.highlightInd !== i) {
      this.setState({ highlightInd: i });
    } else {
      this.setState({ highlightInd: null })
    }
  }

  renderMessage() {
    // console.log(this.props.collections)
    let message;
    if (this.props.collections.length === 0) {
      message = `You don't currently have any new collections ... would you like to make one? You can enter a name below.`;
    } else {
      message = "Select a collection to add the article to, or make a new collection."
    }

    return (
      <p className="thread-text" style={{ color: 'black' }}>
        {message}
      </p>
    )
  }

  renderSelectors() {
    // console.log(this.props.user)

    let allCollections = this.combineAllCollections();

    if (allCollections.length > 0) {

      return (
        <div className="center-all-col">
          {allCollections.map((collection, i) =>
            <Button key={i}
              color="link" className="nav-link"
              style={{ fontSize: '14px', backgroundColor: i === this.state.highlightInd ? '#CADEE8' : 'white' }}
              onClick={() => this.highlight(i)}>{`${collection.name} ( ${collection.articles.length} )`}</Button>)}
        </div>
      )
    } else {
      return null;
    }
  }

  renderInput() {

    return (
      <Form onSubmit={this.handleSubmit}>

        <InputGroup>

          <p style={styles.text}>
            new collection:
            </p>
          <Input style={{ margin: '10px', borderRadius: '25px' }}
            placeholder={" Ex - Cancer Biology "}
            value={this.state.name}
            onChange={(e) => this.handleChange(e.target.value)}
          />
        </InputGroup>

      </Form>
    )
  }

  renderArticlePreview() {
    let title = this.props.article.title;
    let text = [`${title}`, 'will be added to the collection.']
    return (
      <div id="article-preview">
        {text.map((str, i) => {
          return <p key={i} style={i === 0 ? styles.articleText : styles.text}>{str}</p>
        })}
      </div>
    )
  }

  // handleChange changes the value of the input
  handleChange(value) {
    // console.log(value)
    this.setState({
      name: value
    });
  }

  // handleSubmit will fire the submitted data back through Search to App and be recorded
  handleSubmit(e) {
    // console.log('collection input submitted')
    e.preventDefault();

    // if we're updating an existing collection
    if (this.state.highlightInd !== null) {

      let allCollections = this.combineAllCollections();
      let selectedCollection = allCollections[this.state.highlightInd];

      // if it has an owner, its a saved collection
      if (selectedCollection.owner) {
        // send a patch request with the modified collection
        return this.patchCollection(selectedCollection)

      } else {
        // the collection is new, not on the server yet
        // handle it with the modifyCollection function in App
        let collectionName = selectedCollection.name;
        return this.props.modifyCollection(this.props.article, collectionName, 1, () => {
          // console.log('article added to collection')
          this.setState({ name: '' });
        })
      }

    }

    // if we're creating a new collection

    // check there is something there and it isn't just whitespace
    if (!this.state.name || this.state.name.trim().length === 0) {
      this.setState({
        name: ''
      })
      return;
    }

    // otherwise, make a new collection and register it with the function drilled from App
    let collection = {
      name: this.state.name,
      articles: [this.props.article]
    }
    this.props.createNewCollection(collection, () => {
      console.log('collection created!');
      this.setState({ name: '' });
    });

    // reset the highlightInd to prevent empty inputs after submits
    this.setState({
      highlightInd: null
    })

  }

  // this will get fired upon submission, if the user selected a previously saved 
  // collection to add the article to
  patchCollection(collection) {
    collection.articles.push(this.props.article);

    let token = JSON.parse(localStorage.getItem('token'));

    if (!token) {
      return console.log('user is not signed in with a token');
    }

    let headers = {
      Authorization: `Bearer ${token}`,
      Accept: 'application/json',
      'Content-Type': 'application/json',
    }

    // Post the collection to save it to the server. Using the currently stored token
    // will assign the user as the owner of the collection in mongodb on the backend
    return saveCollection(collection, headers, true)
      .then((response) => {
        console.log(response)

        // once we have successfully posted, we will: 

        // 1. Refresh the user's collections - this will send the new collection list sent back
        // from the server to the root App component to update all concerned components
        this.props.refreshUserCollections(response);

      }).catch((e) => {
        console.log(e)
        this.setState({
          error: e
        })
      })
  }

  closeModal() {
    this.props.toggle();
    // set a timeout on the reset here so the content doesn't jump before the modal
    // fades
    setTimeout(() => {
      this.setState({
        highlightInd: null
      })
    }, 1000)
  }

  render() {

    return (
      <Modal centered={true} isOpen={this.props.isVisible} toggle={this.props.toggle}>

        {/* <ModalHeader toggle={this.props.toggle}>Your Collections</ModalHeader> */}
        <ModalBody>
          {this.renderMessage()}

          {/* select an exisiting collection */}
          {this.renderSelectors()}

          {this.renderArticlePreview()}

          {/* name a new one */}
          {this.renderInput()}

        </ModalBody>
        <ModalFooter>
          <Button color="primary" size="sm" onClick={this.handleSubmit}>add article</Button>
          <Button color="secondary" size="sm" onClick={() => this.closeModal()}>close</Button>
        </ModalFooter>

      </Modal>

    );
  }
}

const styles = {
  content: {
    height: '100%',
  },
  text: {
    color: 'black',
    margin: '10px 20px',
    fontSize: '14px'

  },
  articleText: {
    color: 'blue',
    margin: '10px 20px',
    fontWeight: 'bold',
    fontSize: '14px'
  }

}

export default CollectionForm;
