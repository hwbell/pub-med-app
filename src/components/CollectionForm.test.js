import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import CollectionForm from './CollectionForm';

{/* <CollectionForm
  article={this.state.selected}
  isVisible={this.state.collectionModal}
  toggle={this.toggleCollectionForm}
  collections={this.props.collections}
  createNewCollection={this.props.createNewCollection}
  modifyCollection={this.props.modifyCollection} /> */}

// stubs for testing calls
const toggleStub = jest.fn();
const createNewStub = jest.fn();
const modifyStub = jest.fn();

// some sample props
const article = {
  "id": "30865231",
  "source": "MED",
  "pmid": "30865231",
  "doi": "10.1093/molbev/msz059",
  "title": "The Genome of Drosophila innubila Reveals Lineage-Specific Patterns of Selection in Immune Genes.",
  "authorString": "Hill T, Koseva BS, Unckless RL.",
  "journalTitle": "Mol Biol Evol",
  "issue": "7",
  "journalVolume": "36",
  "pubYear": "2019",
  "journalIssn": "0737-4038; 1537-1719; ",
  "pageInfo": "1405-1417",
  "pubType": "journal article",
  "isOpenAccess": "N",
  "inEPMC": "N",
  "inPMC": "N",
  "hasPDF": "N",
  "hasBook": "N",
  "citedByCount": 0,
  "hasReferences": "N",
  "hasTextMinedTerms": "Y",
  "hasDbCrossReferences": "Y",
  "dbCrossReferenceList": {
    "dbName": [
      "EMBL"
    ]
  },
  "hasLabsLinks": "Y",
  "hasTMAccessionNumbers": "N",
  "firstIndexDate": "2019-03-14",
  "firstPublicationDate": "2019-07-01"
}

const collections = [0,1,2,3,4,5].map((num) => {
  return {
    name: `collection ${num+1}`,
    articles: [
      article
    ]
  }
})

const user = {
  name: 'dude',
  collections: collections.slice(3)
}

describe('CollectionForm', () => {

  let wrapper, someProps;
  beforeEach(() => {
    someProps = {
      user,
      article,
      collections,
      isVisible: false,
      toggle: toggleStub,
      createNewCollection: createNewStub,
      modifyCollection: modifyStub
    }

    wrapper = shallow(<CollectionForm {...someProps} />);    
  })

  it('renders without crashing', async () => {
    wrapper.update();
  });
  
  it('renders correctly', async () => {
    const tree = await renderer
      .create(<CollectionForm {...someProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  
  it('contains the correct elements', () => {
    const wrapper = shallow(<CollectionForm {...someProps} />);
    expect(wrapper.find('Modal').length).toEqual(1)
  
    // title 
    expect(wrapper.find('.thread-text').length).toBe(1);
  
    // should have two .nav-link buttons with the names of the provided collections for 
    // both newly created collections and server collections
    expect(wrapper.find('.nav-link').length).toEqual(9);
    expect(wrapper.find('.nav-link').at(0).html()).toContain(collections[0].name);
    expect(wrapper.find('.nav-link').at(1).html()).toContain(collections[1].name);
  
    // two size="sm" buttons - these don't have a unique class right now
    expect(wrapper.find({ size: 'sm' }).length).toEqual(2);
    expect(wrapper.find({ size: 'sm' }).at(0).html()).toContain('add article');
    expect(wrapper.find({ size: 'sm' }).at(1).html()).toContain('close');
  
    // a preview of the article, really its just the title
    expect(wrapper.find('#article-preview').length).toEqual(1);
    expect(wrapper.find('#article-preview').text()).toEqual(article.title+"will be added to the collection.");
  
  })
  
  it('should register localStorage collections if there are none in props', async () => {
    // set the localStorage collections and toss the props collections
    localStorage.setItem('collections', JSON.stringify(collections))
    someProps.collections = [];

    const deepWrapper = mount(<CollectionForm {...someProps} />);
    // this.props.createNewCollection should be called for each of the local collections
    expect(createNewStub.mock.calls.length).toEqual(6);
    
  })
  
  it('should change the title based on whether there are collections or not', async () => {
  
    // with collections
    expect(wrapper.find('.thread-text').at(0).text()).toEqual(
      "Select a collection to add the article to, or make a new collection."
    );
    // without collections
    wrapper.setProps({ collections: [] });
    wrapper.update();
    expect(wrapper.find('.thread-text').at(0).text()).toEqual(
      `You don't currently have any new collections ... would you like to make one? You can enter a name below.`
    );
  })
  
  it('should open and close depending on isVisble prop', async () => {
    const deepWrapper = mount(<CollectionForm {...someProps} />);
  
    // starts out closed
    expect(deepWrapper.find('Modal').props().isOpen).toEqual(false);
  
    // the Modal's isOpen prop changes when different isVisible booleans are provided
    // to the parent
    [true, false, true].forEach((bool) => {
      deepWrapper.setProps({ isVisible: bool })
      deepWrapper.update();
  
      // now should match provided boolean
      expect(deepWrapper.find('Modal').props().isOpen).toEqual(bool);
    })
  
    deepWrapper.setProps({ isVisible: true })
    deepWrapper.update();
  
    // now should be open
    expect(deepWrapper.find('Modal').props().isOpen).toEqual(true);
  
    deepWrapper.setProps({ isVisible: false })
    deepWrapper.update();
  
    // and now closed again
    expect(deepWrapper.find('Modal').props().isOpen).toEqual(false);
  
  })
  
  it('should set the highlightInd for selector buttons', () => {
    const wrapper = shallow(<CollectionForm {...someProps} />);
  
    // the highlightInd should initially be null
    expect(wrapper.state().highlightInd).toEqual(null);
    expect(wrapper.find('.nav-link').length).toEqual(9);
  
    // click the collections and check the highlightInd matches
    wrapper.find('.nav-link').forEach((link, i) => {
      link.simulate('click');
      wrapper.update();
      expect(wrapper.state().highlightInd).toEqual(i);
    })
  
  })
  
  it('should close when "close" button is clicked', () => {
    const deepWrapper = shallow(<CollectionForm {...someProps} />);
  
    // set the highlightInd to a non-null number
    deepWrapper.setState({highlightInd: 1});
    expect(deepWrapper.state().highlightInd).toEqual(1);
  
    // simulate the click
    deepWrapper.find({size: 'sm'}).at(1).simulate('click');
    deepWrapper.update();
  
    // after the call the highlightInd should be set to null, and the toggleStub should
    // have been called. there's a setTimeout on the highlighInd reset, so account for that.
    expect(toggleStub.mock.calls.length).toEqual(1);
  
    setTimeout(() => {
      expect(deepWrapper.state().highlightInd).toEqual(null);
    }, 1100) // the timer in the component is 1000
  })
  
  it('should fire the submit function when "add to collection" button is clicked', () => {
    const deepWrapper = shallow(<CollectionForm {...someProps} />);
  
    // set the highlightInd to a non-empty string
    deepWrapper.setState({name: 'Cancer Biology'});
    expect(deepWrapper.state().name).toEqual("Cancer Biology");
  
    // simulate the click
    deepWrapper.find({size: 'sm'}).at(0).simulate('click', {
      preventDefault: () => {}
    });
    deepWrapper.update();
  
    // after the call the name should be set to an empty string, and the createNewStub should
    // have been called
    expect(deepWrapper.state().highlightInd).toEqual(null);
  
  })

})
