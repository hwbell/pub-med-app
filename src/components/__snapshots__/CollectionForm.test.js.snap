// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionForm renders correctly 1`] = `null`;
